File: typescript_challenges/02_retry.ts
/**
 * Challenge 02 — Async Retry with Exponential Backoff
 * -----------------------------------------------------
 * Implement an async retry utility.
 *
 * async function retry<T>(fn: () => Promise<T>, options?: RetryOptions): Promise<T>
 *
 * interface RetryOptions {
 *   maxAttempts?: number;                          // default 3
 *   baseDelayMs?: number;                          // default 100
 *   shouldRetry?: (err: unknown) => boolean;       // default: always retry
 * }
 *
 * Backoff formula: baseDelayMs * 2^attempt  (attempt is 0-indexed, so the first
 * call has no preceding delay, the first retry waits baseDelayMs, the second
 * retry waits 2*baseDelayMs, etc.)
 *
 * Rules:
 *   - After all attempts are exhausted, reject with the last error.
 *   - If shouldRetry(err) returns false, reject immediately with that error.
 *   - The original call counts as attempt 0; retries start at attempt 1.
 *
 * Usage:
 *   const result = await retry(() => fetchData(url), { maxAttempts: 5 });
 */
export interface RetryOptions {
  maxAttempts?: number;
  baseDelayMs?: number;
  shouldRetry?: (err: unknown) => boolean;
}

export async function retry<T>(
  fn: () => Promise<T>,
  options?: RetryOptions
): Promise<T> {
  throw new Error("not implemented");
}

File: go_challenges/02_retry/02_retry.go
// Challenge 02 — Retry with Exponential Backoff
// -----------------------------------------------
// Implement Retry[T any] that re-runs a fallible function with exponential backoff.
//
//	func Retry[T any](fn func() (T, error), opts ...RetryOption) (T, error)
//
// Default behaviour (no options):
//   - Max 3 attempts
//   - Base delay 100 ms, doubles each retry: 100 ms, 200 ms, 400 ms, …
//   - Retries on any non-nil error
//   - Returns the zero value and the last error after all attempts are exhausted
//
// Options (functional-options pattern):
//
//	WithMaxAttempts(n int) RetryOption
//	WithBaseDelay(d time.Duration) RetryOption
//	WithRetryOn(pred func(error) bool) RetryOption
//	    Only retry when pred(err) returns true; propagate other errors immediately.
//
// Usage:
//
//	result, err := Retry(callAPI,
//	    WithMaxAttempts(5),
//	    WithBaseDelay(50*time.Millisecond),
//	)

package main

import (
	"time"
)

// RetryOption is a functional option for Retry.
type RetryOption func(*retryConfig)

type retryConfig struct {
	// your fields here
}

func WithMaxAttempts(n int) RetryOption {
	panic("not implemented")
}

func WithBaseDelay(d time.Duration) RetryOption {
	panic("not implemented")
}

func WithRetryOn(pred func(error) bool) RetryOption {
	panic("not implemented")
}

// Retry calls fn up to the configured number of times, sleeping between attempts.
func Retry[T any](fn func() (T, error), opts ...RetryOption) (T, error) {
	panic("not implemented")
}

var _ = time.Millisecond

func main() {
	attempts := 0
	result, err := Retry(func() (string, error) {
		attempts++
		if attempts < 3 {
			return "", fmt.Errorf("transient error")
		}
		return "success", nil
	}, WithMaxAttempts(5))
	fmt.Printf("result: %v, err: %v, attempts: %d\n", result, err, attempts)
}
